sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,347 @@
+            var propertySymbol = symbol as IPropertySymbol;
+
+            if (propertySymbol == null)
+            {
+                return FindOriginatingSymbol(symbol);
+            }
+
+            return FindOriginatingSymbol(idRef.Parent.Parent is AssignmentExpressionSyntax
+                    ? propertySymbol.SetMethod
+                    : propertySymbol.GetMethod);
+        }
+
+        private ITypeSymbol HandleInvocation(IdentifierNameSyntax invokedOn, ISymbol invocationSymbol,
+            SemanticModel semanticModel)
+        {
+            var methodSymbol = invocationSymbol as IMethodSymbol;
+            if (methodSymbol == null)
+            {
+                return null;
+            }
+
+            return methodSymbol.IsExtensionMethod
+                ? semanticModel.GetTypeInfo(invokedOn).ConvertedType
+                : FindOriginatingSymbol(invocationSymbol);
+        }
+
+        private INamedTypeSymbol FindOriginatingSymbol(ISymbol accessedMember)
+        {
+            if (accessedMember == null)
+            {
+                return null;
+            }
+
+            var originatingInterface = accessedMember.ContainingType
+                   .AllInterfaces
+                   .Where(i => HasImplementationForMember(i, accessedMember))
+                   .FirstOrDefault();
+
+            if (originatingInterface != null)
+            {
+                return originatingInterface;
+            }
+
+            var overridenSymbol = SymbolHelper.GetOverriddenMember(accessedMember);
+            return overridenSymbol != null
+                ? overridenSymbol.ContainingType
+                : accessedMember.ContainingType;
+        }
+
+        private static bool HasImplementationForMember(INamedTypeSymbol namedSymbol, ISymbol accessedMember)
+        {
+            return namedSymbol
+                .GetMembers()
+                .Any(member =>
+                    accessedMember.Equals(
+                        accessedMember.ContainingType.FindImplementationForInterfaceMember(member)));
+        }
+
+        private class ParameterData
+        {
+            public bool ShouldReportOn { get; set; } = true;
+
+            private readonly IParameterSymbol parameterSymbol;
+            private readonly HashSet<ITypeSymbol> usedAs = new HashSet<ITypeSymbol>();
+
+            public ParameterData(IParameterSymbol parameterSymbol)
+            {
+                this.parameterSymbol = parameterSymbol;
+            }
+
+            public void AddUsage(ITypeSymbol symbolUsedAs)
+            {
+                usedAs.Add(symbolUsedAs);
+            }
+
+            public bool MatchesIdentifier(IdentifierNameSyntax id, SemanticModel semanticModel)
+            {
+                var symbol = semanticModel.GetSymbolInfo(id).Symbol;
+                return Equals(parameterSymbol, symbol);
+            }
+
+            public Diagnostic GetRuleViolation()
+            {
+                if (ShouldReportOn)
```

Comment:
I know it will add one more `return null` but I would simplify this nesting by inverting the condition here.